### PR TITLE
Continue revision of ranged weapons [WIP]

### PR DESCRIPTION
**Pull request by mugling**
Follows #17656
- MIN_RECOIL macro becomes `extern` variable to aid future adjustment
- Initial implementation of `player::recoil_current`
- Turrets and NPC's use `engagement::effective_min`
- `dispersion` now follows normal distribution (fluke hit/misses are possible but improbable)

### CODE REVIEW

File: src/ranged.cpp [hunk 1]
```diff
@@ -119,7 +119,7 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
     // for the isosceles triangle formed by the intended and actual targets
     // we can use the cosine formula (a² = b² + c² - 2bc⋅cosθ) to calculate the tangent
     double range = rl_dist( source, target_arg );
-    double missed_by = sqrt( 2 * pow( range, 2 ) - ( 2 * pow( range, 2 ) * cos( ARCMIN( dispersion ) ) ) );
+    double missed_by = sqrt( 2 * ( pow( range, 2 ) ) * ( 1 - cos( ARCMIN( dispersion ) ) ) );
```
Comment (codemime) [inline src/ranged.cpp]: The brackets around `pow()` are excessive.

**Comment by Coolthulhu** [comment]
If you're changing turrets, you should also fix #17343.

**Comment by mugling** [comment]
> If you're changing turrets, you should also fix #17343.

Going to tackle turrets as a whole in a separate PR

**Comment by mugling** [comment]
Ready for review

File: src/npcmove.cpp [hunk 2]
```diff
@@ -1160,13 +1160,7 @@ int npc::confident_gun_mode_range( const item::gun_mode &gun, int at_recoil ) co
...
     }
 
-    double deviation = get_weapon_dispersion( gun.target, false ) + at_recoil;
-    // Halve to get expected values
-    deviation /= 2;
-    // Convert from MoA back to quarter-degrees.
-    deviation /= 15;
-
-    int ret = std::min( int( confidence_mult() * 360 / deviation ), gun->gun_range( this ) );
+    int ret = gun_engagement_range( *gun, engagement::effective_min, at_recoil );
```
Comment (Coolthulhu) [inline src/npcmove.cpp]: Must take confidence requirement into account.

Comment (mugling) [inline src/npcmove.cpp]: Is the range for that `[0.25-1.25]`?

**Comment by Coolthulhu** [comment]
I'll need a comparison of sample values using old formula and new formula.
At the very least it must be expected values for 3 use cases, one of which should be have as many "effects" as possible (ie. penalty for encumbrance, for recoil, for low skill etc.).

**Comment by mugling** [comment]
> I'll need a comparison of sample values using old formula and new formula.

Which formula?

File: src/npcmove.cpp (hunk 2 above)
Comment (Coolthulhu) [inline src/npcmove.cpp]: Yes.
You can change the function to return engagement ranges instead, but then there should be a new engagement type: 100% guaranteed good hit at full aim (for strictly precise setting).

**Comment by Coolthulhu** [comment]
> Which formula?

Final dispersion value.

`missed_by` differs from the before-revision values mostly linearly, but dispersion probably won't.
Expected value doesn't say the whole story - it is the bare minimum for proper review.

The best statistic would be a bunch of charts with `y` axis being probability, `x` being distance, with headshot, crit, good hit, glance and miss probabilities being overlayed on each other.
There should be at least 3 such charts, for each of the sample characters: "sniper" (8 skill, 5 encumbrance, 45 recoil), "tank" (5 skill, 30 encumbrance, 300 recoil) and "rookie" (2 skill, 15 encumbrance, 90 recoil).

**Comment by mugling** [comment]
> Final dispersion value.

As in the output of `get_weapon_dispersion`?

**Comment by Coolthulhu** [comment]
> As in the output of get_weapon_dispersion?

Whatever gets passed to `Creature::projectile_attack`, assuming no burst.

**Comment by mugling** [comment]
`missed_by` remains in linear proportion to total `dispersion`?

The graphs you want would be the same both before this PR and after?

**Comment by Coolthulhu** [comment]
> missed_by remains in linear proportion to total dispersion?

Roughly. The old approximation was surprisingly accurate - for pretty much all sane values of `dispersion` and range, the difference was just around 10%.

> The graphs you want would be the same both before this PR and after?

Yes. They can be calculated by hand or brute forced, as long as they're accurate enough.

**Comment by mugling** [comment]
> The graphs you want would be the same both before this PR and after?
 
> > Yes.

I'm still lost. If nothing has changed why do we want graphs. If something has changed what are we graphing?

**Comment by Coolthulhu** [comment]
> I'm still lost. If nothing has changed why do we want graphs. If something has changed what are we graphing?

I thought you were asking what should the graphs show.

They shouldn't show the same results, just the same thing: dispersion rolls.

**Comment by mugling** [comment]
> They shouldn't show the same results, just the same thing: dispersion rolls.

After this PR they are normally distributed about the mean. Is there much value in graphing that given it's a lot of work to look at a standard curve?

I did some preliminary testing detailed in the comments of #17656 - before this PR `dispersion` ranges form [0..max] biased towards 57-72% of max dependent upon recoil.

Overall balance isn't adversely affected when you consider that #17656 made aiming about 10% harder and `recoil` is currently a little too punishing anyway.

**Comment by Coolthulhu** [comment]
> After this PR they are normally distributed about the mean. Is there much value in graphing that given it's a lot of work to look at a standard curve?

You can table some sample results, but I'll need a good way to see where is it all going.

**Comment by mugling** [comment]
`--dump-stats GUN HTML > gun.html` can provide automated output of tabular data from which unit tests can be constructed. What if any additional columns do you require?

**Comment by Coolthulhu** [comment]
> What if any additional columns do you require?

Not sure how you want to format it, but I'll need those:
* Chance to headshot/crit/miss/etc. tabled by range
* Effective ranges for at least 3 guns for the 3 characters I described above
* Same thing for old formulas

**Comment by mugling** [comment]
> Chance to headshot/crit/miss/etc. tabled by range

That's a different approach to the one outlined in #17643 which was gradual revision with unit tests and example data via the stats dumper.

It's not feasible to reproduce that data for each sequential PR. Unit tests are much better as we can check assumptions remain valid and there are no regressions.

> Whatever gets passed to Creature::projectile_attack

Calculated weapon `dispersion` is an intermediate value. It would be better to test on `missed_by` or `range` as those are end results. A unit test *'shotgun has an effective range of 5'* is more descriptive and less brittle than *'shotgun has a nett dispersion of 300'* as the latter is heavily dependent upon what `Creature::projectile_attack` does with `dispersion`.

> Effective ranges for at least 3 guns for the 3 characters I described above

These can become columns for the stats dumper

> Same thing for old formulas

To be clear thought the only change this PR should result in is that total calculated dispersion is now a normal distribution as opposed to a somewhat right-skewed uniform distribution. I don't see the value in graphing the former especially as I already included an approximation of the skewness and graphing a normal distribution achieves nothing.

If you disagree that it should be a normal distribution we should consider that more but I'm not sure what your otherwise trying to get at and what your asking for is a lot of ongoing work on my part and not part of the earlier discussion.

**Comment by Coolthulhu** [comment]
> A unit test 'shotgun has an effective range of 5' is more descriptive and less brittle than 'shotgun has a nett dispersion of 300' as the latter is heavily dependent upon what Creature::projectile_attack does with dispersion.

I will require a dispersion value for comparison.
Effective range alone may be (likely is) hiding some important parts of the results. 

> To be clear thought the only change this PR should result in is that total calculated dispersion is now a normal distribution as opposed to a somewhat right-skewed uniform distribution. I don't see the value in graphing the former especially as I already included an approximation of the skewness and graphing a normal distribution achieves nothing.

At least table it then. Or table an approximation with a note that it should be roughly this much.

> If you disagree that it should be a normal distribution we should consider that more but I'm not sure what your otherwise trying to get at and what your asking for is a lot of ongoing work on my part and not part of the earlier discussion.

Normal distribution is fine.
The problem is, you're trying to get a major change in, but are trying to avoid all the work that is necessary to evaluate the change.
Dumping a table of sample results isn't "a lot of work". When I tested your `missed_by` formula PR, I wrote a quick script in python just to compare the results.

It doesn't need to be a fancy graph, but I can't rate a major balancing PR based on "it feels kinda OK for those 2 cases, I guess" or "effective range, which highly simplifies the whole formula, is roughly the same".

**Comment by mugling** [comment]
> Effective range alone may be (likely is) hiding some important parts of the results.

What might it hide given that the relationship between `missed_by`, `range` and `dispersion` is linear?

No sarcasm intended - you correctly identified my earlier assumption that the sum of uniform distributions was itself uniform as an error so there may be something I haven't considered?

> the only change this PR should result in is that total calculated dispersion is now a normal distribution
> > At least table it then.

You want a table of a normal distribution? Can you give a terse example of what output you want?

> but are trying to avoid all the work that is necessary to evaluate the change.

That's a little unfair. I gave an extensive worked example of the problem in #17643 complete with code references, then started work on the requested units tests and stats output in #17656.

> Dumping a table of sample results isn't "a lot of work".

A series of graphs is a lot of work, especially if it needs repeating for each PR. It's also more work than I anticipated - my argument is the goal posts have now been moved. That's fine providing you can make a clear case for what extra data you need and why. I'm not sure on either of those at the moment.

**Comment by Coolthulhu** [comment]
> What might it hide given that the relationship between missed_by, range and dispersion is linear?

The distribution.
It is calculated in a different way than brute forcing `n` rolls and averaging the results, meaning there is a high chance it skips some important value, for example by not adding the mean where it should be added or something like that.

> That's a little unfair. I gave an extensive worked example of the problem in #17643 complete with code references, then started work on the requested units tests and stats output in #17656.

I take back the "all" part.
Still, the comparison of the actual values is important and can't be skipped over in favor of an arbitrary measure that is calculated separately from the important ones and only correlates with the actual results.

> A series of graphs is a lot of work, especially if it needs repeating for each PR.

Then just show a table of the results, like I said when you first complained about graphs.
Graphs would be better, but a table can suffice if it is done right.


Also, I don't trust that `gun_engagement_range` formula. An unit test that ensures it calculates what it is trying to calculate would be useful. Something like:
* Find engagement range
* Make 100 random rolls
* Check that the proportion of hits (of required "grade") to misses (or hits too weak to count) is roughly the one we asked for in the engagement range argument

**Comment by mugling** [comment]
> An unit test that ensures it calculates what it is trying to calculate would be useful

How to make the unit test deterministic for any practical number of iterations?

**Comment by Coolthulhu** [comment]
> How to make the unit test deterministic for any practical number of iterations?

Doesn't need to be, `monster_test.cpp` has a non-deterministic test already.
Make sure it passes more often that in doesn't and then bump the number of iterations 10 times and it should be working.

**Comment by mugling** [comment]
> Also, I don't trust that gun_engagement_range formula

See pushed unit test

File: src/ranged.cpp [hunk 3]
```diff
@@ -549,7 +549,7 @@ int player::fire_gun( const tripoint &target, int shots, item& gun )
             break;
         }
 
-        double dispersion = get_weapon_dispersion( &gun, true );
+        double dispersion = rng_normal( get_weapon_dispersion( gun ) ) + rng_normal( recoil_current() );
```
Comment (kevingranade) [inline src/ranged.cpp]: What part of, "this is working as intended, don't change it" was unclear?

Comment (kevingranade) [inline src/ranged.cpp]: At a minimum, we need the data Coolthulu is asking for, if I'm wrong and they're very close to each other, great, but I strongly doubt it.
Alternately, you can extract the random calls to one place, maintaining the separate sources of dispersion until it is applied.

Comment (mugling) [inline src/ranged.cpp]: > What part of, "this is working as intended, don't change it" was unclear?

I don't think the current code as a whole works well and that was the argument set out in #17643. The sum of distributions part isn't especially causative but it is a significant obstacle (see below).

> you can extract the random calls to one place, maintaining the separate sources of dispersion until it is applied.

This PR starts work on that idea - it breaks it out into player recoil and weapon parts. At `MIN_RECOIL` there isn't a significant difference (normal distribution is unbiased, original bias was 57%). At higher recoil there is (maximum tested was 72%).

The rationale is that it's impossible for `gun_engagement_range` to perform the inverse calculation for a sum of distributions due to the variable bias.

> Normal distribution is fine.

I'd agree a limited set of summed normal distributions is likely the correct model. The next steps could include working out what that set is and providing some data to check the output is sensible and writing some unit tests.

**Comment by mugling** [comment]
Suppose player `recoil` is 200 and `get_weapon_dispersion()` returns 300:

| value        | before PR | after PR |
|--------------|----------|-------|
| recoil        | 200      | 200  |
| dispersion | 300     | 300  |
| range        | 0-300 | 0-300 |
| probability density   | Irwin–Hall? | normal  |
| average | ? | 150 |

If we increase `recoil` to 300 then `get_weapon_dispersion()` returns 440:

| value        | before PR | after PR |
|--------------|----------|-------|
| recoil        | 300      | 300  |
| dispersion | 450     | 450  |
| range        | 0-450 | 0-450 |
| probability density   | Irwin–Hall? | normal  |
| average | ? | 220 |

The only way to calculate the **average** dispersion before this PR is to know the entire workings of  `get_weapon_dispersion()`. Suppose we did this for some arbitrary values: in both variants `recoil` affects the `dispersion` by the same amount but in the latter it also affects the **average**.

| recoil | before PR | after PR |
|---------|--------------|-----------|
| 100    | 48%          | 50%      |
| 150    | 52%          | 50%      |
| 200    | 58%          | 50%      |
| 250    | 54%          | 50%      |

What is the percentage for a `recoil` of 210? Is it `50%`, `54%` or perhaps `44%`? Now add some more variables (encumbrance, skills, stats)... 

Having to model the entire implementation in a test-case isn't a unit test. Such problems are why Irwin-Hall is a particularly nasty distribution to used compared to normal.

I'm not objecting to providing more graph data or unit tests but this is only feasible if it's possible if we use a probability distribution function that doesn't depend upon close to a dozen variables.

**Comment by mugling** [comment]
More succinctly after this PR your graphics would be 
`recoil vs average dispersion`
`encumbrance vs average dispersion`

Before it would be:
`recoil vs average dispersion at encumbrance 1`
`recoil vs average dispersion at encumbrance 2`
`recoil vs average dispersion at encumbrance 3`
...
`recoil vs average dispersion at encumbrance n`

**Comment by Coolthulhu** [comment]
Unit tests confirming similarity of the new and old systems aren't needed - after all, we'll be throwing the old one out.

> I'm not objecting to providing more graph data or unit tests but this is only feasible if it's possible if we use a probability distribution function that doesn't depend upon close to a dozen variables.

Don't compare entire distributions of variables, compare distributions of scenarios. Like the ones I provided as an example.

For example, the table can have columns like:
* Sniper (old system)
* Sniper (new system)
* Rookie (old system)

etc.
The rows would be:
* Effective ranges (without rounding) for selected accuracies
* Chance of headshot/crit/etc. at full aim and at no aim
* (optional) Average damage multiplier on attacks, counting miss as 0%

**Comment by mugling** [comment]
Raw data showing `missed_by` at `range`:

| Name     | Ammo | 1    | 2    | 3    | 4    | 5    | 6    | 7    | 8    | 9    | 10   |
|----------|------|------|------|------|------|------|------|------|------|------|------|
| AR-15    | 223  | 0.16 | 0.31 | 0.47 | 0.62 | 0.78 | 0.93 | 1.09 | 1.25 | 1.4  | 1.56 |
| Glock 19 | 9mm  | 0.2  | 0.4  | 0.59 | 0.79 | 0.99 | 1.19 | 1.39 | 1.58 | 1.78 | 1.98 |

**Comment by mugling** [comment]
`range` vs `missed_by` was linear and remains so:

d2

**Comment by mugling** [comment]
`turns spent aiming` versus `range` where AR-15+ includes `rifle scope`

d4

This is much more interesting and highlights @kevingranade's earlier point that aiming is currently insignificant.

How about extending the dump tool to export the above as a gnuplot for each of the scenarios you proposed? We then know we have the correct implementation when those graphs look correct.

**Comment by mugling** [comment]
As to the above?

**Comment by mugling** [comment]
As to the above?

**Pull request closed without merging**